oracle/metrics/dynamic_metrics.go

```diff
@@ -0,0 +1,121 @@
+package metrics
+
+import (
+	"context"
+	"time"
+
+	"github.com/skip-mev/connect/v2/oracle/config"
+)
+
+type ImplType int
+
+const (
+	OracleMetricsType ImplType = iota
+	NoOpMetricsType
+	UnknownMetricsType
+)
+
+// determines the underlying impl of the metrics.
+func determineMetricsType(m Metrics) ImplType {
+	switch m.(type) {
+	case *OracleMetricsImpl:
+		return OracleMetricsType
+	case *noOpOracleMetrics:
+		return NoOpMetricsType
+	default:
+		return UnknownMetricsType
+	}
+}
+
+var _ Metrics = &dynamicMetrics{}
+
+// dynamicMetrics is a type that can change its internal metrics impl on the fly.
+// this is useful for when a connect instance is started before a node. we can't be sure which one starts first,
+// so we need to be able to switch when the node comes online.
+type dynamicMetrics struct {
+	cfg         config.MetricsConfig
+	nc          NodeClient
+	impl        Metrics
+	metricsType ImplType
+}
+
+func NewDynamicMetrics(ctx context.Context, cfg config.MetricsConfig, nc NodeClient) Metrics {
+	impl := NewMetricsFromConfig(cfg, nc)
+	dyn := &dynamicMetrics{
+		cfg:         cfg,
+		nc:          nc,
+		impl:        impl,
+		metricsType: determineMetricsType(impl),
+	}
+	// we only want to kick off the routine of attempting to switch if we're a noop metrics, telemetry is enabled,
+	// and we have a node client.
+	if dyn.metricsType == NoOpMetricsType && !cfg.Telemetry.Disabled && nc != nil {
+		dyn.retrySwitchImpl(ctx)
+	}
+	return dyn
+}
+
+// retrySwitchImpl kicks off a go routine that attempts to contact a node every 3 seconds for 10 mins.
+// if it gets a response, it will switch its internal metrics impl.
+func (d *dynamicMetrics) retrySwitchImpl(ctx context.Context) {
+	go func() {
+		retryDuration := time.NewTimer(10 * time.Minute)
+		ticker := time.NewTicker(3 * time.Second)
+
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-retryDuration.C:
+				return
+			case <-ticker.C:
+				// this is technically a race condition, but it doesn't really matter. if something is accessing the
```

Comment:
What do you mean? Are you saying this is a race condition because the oracle would be using `d.impl` for reporting metric while we do this change here?

Reply:
correct